README.md
# OnBoarding Animation

<br />

OnBoarding Animation provides page like animation to the onBoarding screens with the page indicator.

<br />

## Screenshots and Screen recording
![OnBoarding Animation](https://github.com/mi-ghanshyam/onboarding_animation/blob/master/assets/onboarding_animation.gif)

<br />

## Usage

#### Installation

Add `onboarding_animation: any` to your `pubspec.yaml` dependencies. And import it:

```dart
import 'package:onboarding_animation/onboarding_animation.dart';
```

#### Example

    OnBoardingAnimation(
              pages: [
               Container(),
               Container(),
               Container(),
              ],
              indicatorDotHeight: 7.0,
              indicatorDotWidth: 7.0,
              indicatorType: IndicatorType.expandingDots,
              indicatorPosition: IndicatorPosition.bottomCenter,
            ),

<br />

## Parameters

### Required Parameters

#### pages:
- Need to pass the List of Widgets for providing the content of onBoarding cards.

<br />

### Optional Parameters

### controller:

- This parameter is use to set the page controller.

#### indicatorOffset:
- This parameter is used to set the offset of the indicator.

#### indicatorDotWidth:
- Set the width of the the indicator's dot.

#### indicatorDotHeight:
- Set the height of indicator's dot.

#### indicatorDotSpacing:
- Using this parameter user can define the space between the dots of indicator.

#### indicatorDotRadius:
- This is use to set the radius of the circle shown in the indicator.

#### indicatorInActiveDotColor:
- This is use to set the color of the inactive(unselected) dot color.

#### indicatorActiveDotColor:
- This is use to set the color of the active(selected) dot color.

#### indicatorStrokeWidth:
- indicatorStrokeWidth is use to set the width of the stroke if the PaintStyle is selected to the stroke.

#### indicatorType:
- This is an enum which is use to select the type of an indicator.

#### indicatorPosition:
- This is an enum which is use to select the position of an indicator.

#### indicatorColorTransitionActiveStrokeWidth:
- Need to pass the List of Widgets for the providing the content of the onBoarding cards.

#### indicatorExpansionFactor:
- indicatorExpansionFactor is multiplied by indicatorDotWidth to resolve the width of the expanded dot.

#### indicatorJumpScale:
- The maximum scale the dot will hit while jumping.

#### indicatorVerticalOffset:
- The vertical offset of the jumping dot.

#### indicatorPaintStyle:
- indicatorPaintStyle is use to select between the fill and the stroke style.

#### indicatorWormType:
- indicatorWormType helps to select the form of the WormType between normal and thin.

#### indicatorSwapType:
- indicatorSwapType helps to select the form of the SwapType between normal, yRotation and zRotation.

#### indicatorActivePaintStyle:
- Inactive dots paint style (fill/stroke) defaults to fill.

#### indicatorActiveStrokeWidth:
- This is ignored if indicatorActivePaintStyle is PaintStyle.fill.

#### indicatorScale:
- indicatorScale is multiplied by indicatorDotWidth to resolve active dot scaling.

#### indicatorActiveDotScale:
- indicatorActiveDotScale is multiplied by indicatorDotWidth to resolve active dot scaling.

#### indicatorMaxVisibleDots:
- The max number of dots to display at a time if count is <= maxVisibleDots indicatorMaxVisibleDots = countmust be an odd number that's >= 5.

#### indicatorFixedCenter:
- If True the old center dot style will be used.

#### indicatorInActiveDotDecoration:
- indicatorInActiveDotDecoration is used to set the inactive dot decoration if indicatorType is selected to custom.

#### indicatorActiveDotDecoration:
- indicatorActiveDotDecoration is used to set the active dot decoration if indicatorType is selected to custom.

#### indicatorActiveColorOverride:
- This parameter is use to override the existing active colors.

#### indicatorInActiveColorOverride:
- This parameter is use to override the existing inactive colors.

<br />

## Guideline for contributors
Contribution towards our repository is always welcome, we request contributors to create a pull request to the development branch only.

<br />

## Guideline to report an issue/feature request
It would be great for us if the reporter can share the below things to understand the root cause of the issue.
- Library version
- Code snippet
- Logs if applicable
- Device specification like (Manufacturer, OS version, etc)
- Screenshot/video with steps to reproduce the issue

<br />

## Library used
- [SmoothPageIndicator](https://pub.dev/packages/smooth_page_indicator "SmoothPageIndicator")

<br />

# LICENSE!
OnBoarding Animation is [MIT-licensed](https://github.com/mi-ghanshyam/onboarding_animation/blob/master/LICENSE "MIT-licensed").

<br />

# Let us know!
We’d be really happy if you send us links to your projects where you use our component. Just send an email to sales@mindinventory.com And do let us know if you have any questions or suggestion regarding our work.


CHANGELOG.md
## [0.0.3] - 2023-04-18

* Provided the accessibility of all the properties of smooth_page_indicator.

## [0.0.2] - 2022-09-07

* Added support for page controller in addition to being able to control the pages.

## [0.0.1] - 2022-01-13

* Initial release.

test/onboarding_animation_test.dart
/*
import 'package:flutter_test/flutter_test.dart';

import 'package:onboarding_animation/onboarding_animation.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}
*/


lib/onboarding_animation.dart
library onboarding_animation;

import 'package:flutter/material.dart';
import 'package:smooth_page_indicator/smooth_page_indicator.dart';
export 'package:smooth_page_indicator/smooth_page_indicator.dart';

class OnBoardingAnimation extends StatefulWidget {
  /// This parameter is required. need to provide list of [Widget] user wants to show on onBoarding.
  final List<Widget> pages;

  /// This parameter is use to set the page controller.
  final PageController? controller;

  /// This parameter is used to set the offset of the indicator.
  final double indicatorOffset;

  /// Set the width of the the indicator's dot.
  final double indicatorDotWidth;

  /// Set the height of indicator's dot.
  final double indicatorDotHeight;

  /// Using this parameter user can define the space between the dots of indicator.
  final double indicatorDotSpacing;

  /// Is use to set the radius of the circle shown in the indicator.
  final double indicatorDotRadius;

  /// This is use to set the color of the inactive(unselected) dot color.
  final Color indicatorInActiveDotColor;

  /// This is use to set the color of the active(selected) dot color.
  final Color indicatorActiveDotColor;

  /// [indicatorStrokeWidth] is use to set the width of the stroke if the [PaintStyle] is selected to the stroke.
  final double indicatorStrokeWidth;

  /// This is an enum which is use to select the type of an indicator.
  final IndicatorType indicatorType;

  /// This is an enum which is use to select the position of an indicator.
  final IndicatorPosition indicatorPosition;

  /// [indicatorExpansionFactor] is multiplied by [indicatorDotWidth] to resolve
  /// the width of the expanded dot.
  final double indicatorExpansionFactor;

  /// The maximum scale the dot will hit while jumping.
  final double indicatorJumpScale;

  /// The vertical offset of the jumping dot.
  final double indicatorVerticalOffset;

  /// [indicatorPaintStyle] is use to select between the fill and the stroke style.
  final PaintingStyle indicatorPaintStyle;

  /// [indicatorWormType] helps to select the form of the [WormType] between normal and thin.
  final WormType indicatorWormType;

  /// [indicatorSwapType] helps to select the form of the [SwapType] between normal, yRotation and zRotation.
  final SwapType indicatorSwapType;

  /// Inactive dots paint style (fill/stroke) defaults to fill.
  final PaintingStyle indicatorActivePaintStyle;

  /// This is ignored if [indicatorActivePaintStyle] is PaintStyle.fill.
  final double indicatorActiveStrokeWidth;

  /// [indicatorScale] is multiplied by [indicatorDotWidth] to resolve
  /// active dot scaling.
  final double indicatorScale;

  /// [indicatorActiveDotScale] is multiplied by [indicatorDotWidth] to resolve
  /// active dot scaling.
  final double indicatorActiveDotScale;

  /// The max number of dots to display at a time
  /// if count is <= [maxVisibleDots] [indicatorMaxVisibleDots] = count
  /// must be an odd number that's >= 5.
  final int indicatorMaxVisibleDots;

  /// If True the old center dot style will be used.
  final bool indicatorFixedCenter;

  /// [indicatorInActiveDotDecoration] is used to set the inactive dot decoration if [indicatorType] is selected to custom.
  final DotDecoration? indicatorInActiveDotDecoration;

  /// [indicatorActiveDotDecoration] is used to set the active dot decoration if [indicatorType] is selected to custom.
  final DotDecoration? indicatorActiveDotDecoration;

  /// This parameter is use to override the existing active and inactive colors.
  final ColorBuilder? indicatorActiveColorOverride,
      indicatorInActiveColorOverride;

  const OnBoardingAnimation({
    required this.pages,
    this.controller,
    this.indicatorOffset = 10.0,
    this.indicatorDotWidth = 10.0,
    this.indicatorDotHeight = 10.0,
    this.indicatorDotSpacing = 8.0,
    this.indicatorDotRadius = 11.0,
    this.indicatorJumpScale = 1.4,
    this.indicatorVerticalOffset = 0.0,
    this.indicatorInActiveDotColor = Colors.grey,
    this.indicatorActiveDotColor = Colors.black,
    this.indicatorPaintStyle = PaintingStyle.fill,
    this.indicatorActivePaintStyle = PaintingStyle.fill,
    this.indicatorScale = 1.4,
    this.indicatorActiveDotScale = 1.3,
    this.indicatorMaxVisibleDots = 5,
    this.indicatorFixedCenter = false,
    this.indicatorActiveStrokeWidth = 1.0,
    this.indicatorExpansionFactor = 3.0,
    this.indicatorStrokeWidth = 1.0,
    this.indicatorWormType = WormType.normal,
    this.indicatorSwapType = SwapType.normal,
    this.indicatorInActiveDotDecoration,
    this.indicatorActiveDotDecoration,
    this.indicatorActiveColorOverride,
    this.indicatorInActiveColorOverride,
    this.indicatorType = IndicatorType.worm,
    this.indicatorPosition = IndicatorPosition.bottomCenter,
    Key? key,
  }) : super(key: key);

  @override
  _OnBoardingAnimationState createState() => _OnBoardingAnimationState();
}

class _OnBoardingAnimationState extends State<OnBoardingAnimation> {
  late PageController _pageController;
  final ValueNotifier<double> _pageIndex = ValueNotifier(0.0);

  /// Initialize the listener to add page listener.
  @override
  void initState() {
    super.initState();
    _pageController = widget.controller ?? PageController(initialPage: 0);
    WidgetsBinding.instance.addPostFrameCallback((_) {
      _pageIndex.value = _pageController.page ?? 0;
      _pageController.addListener(_listener);
    });
  }

  /// For disposing the PageController and removing the page listener.
  @override
  void dispose() {
    super.dispose();
    _pageController.removeListener(_listener);
    _pageController.dispose();
  }

  /// Main body for onBoarding animation.
  @override
  Widget build(BuildContext context) {
    return SafeArea(
      child: Stack(
        alignment: Alignment.bottomCenter,
        children: [
          ValueListenableBuilder<double>(
            builder: (_, indexValue, __) {
              return PageView.builder(
                itemCount: widget.pages.length,
                controller: _pageController,
                itemBuilder: (BuildContext context, int index) {
                  /// TweenAnimationBuilder to smoothen the animation.
                  return TweenAnimationBuilder(
                    duration: const Duration(milliseconds: 140),
                    tween: Tween<double>(
                      begin: ((((index - indexValue) * 100.0) *
                              ((index - indexValue).abs() / 40)) /
                          100.0),
                      end: ((((index - indexValue) * 100.0) *
                              ((index - indexValue).abs() / 40)) /
                          100.0),
                    ),
                    builder: (_, double rotation, _child) {
                      return Transform(
                        transform: Matrix4.identity()
                          ..setEntry(3, 2, .1)
                          ..rotateY(rotation),
                        alignment: FractionalOffset.center,
                        child: widget.pages[index],
                      );
                    },
                  );
                },
                // onPageChanged: (index) => _pageIndex.value = index.toDouble(),
              );
            },
            valueListenable: _pageIndex,
          ),
          Align(
            alignment: _getIndicatorPosition()!,
            child: SmoothPageIndicator(
              controller: _pageController,
              count: widget.pages.length,
              effect: _getIndicatorType(),
            ),
          )
        ],
      ),
    );
  }

  /// This method listen to the page changes to give the position value of the pages using [PageController].
  void _listener() {
    _pageIndex.value = _pageController.page!;
  }

  /// This method is used to set the position of the page indicator.
  Alignment? _getIndicatorPosition() {
    switch (widget.indicatorPosition) {
      case IndicatorPosition.center:
        return Alignment.center;
      case IndicatorPosition.topCenter:
        return Alignment.topCenter;
      case IndicatorPosition.topRight:
        return Alignment.topRight;
      case IndicatorPosition.topLeft:
        return Alignment.topLeft;
      case IndicatorPosition.bottomCenter:
        return Alignment.bottomCenter;
      case IndicatorPosition.bottomRight:
        return Alignment.bottomRight;
      case IndicatorPosition.bottomLeft:
        return Alignment.bottomLeft;
      case IndicatorPosition.centerRight:
        return Alignment.centerRight;
      case IndicatorPosition.centerLeft:
        return Alignment.centerLeft;
    }
  }

  /// This method is use to select the type of indicator from the [SmoothPageIndicator].
  dynamic _getIndicatorType() {
    switch (widget.indicatorType) {
      case IndicatorType.worm:
        return WormEffect(
          paintStyle: widget.indicatorPaintStyle,
          strokeWidth: widget.indicatorStrokeWidth,
          dotColor: widget.indicatorInActiveDotColor,
          type: widget.indicatorWormType,
          dotWidth: widget.indicatorDotWidth,
          dotHeight: widget.indicatorDotHeight,
          activeDotColor: widget.indicatorActiveDotColor,
          offset: widget.indicatorOffset,
          radius: widget.indicatorDotRadius,
          spacing: widget.indicatorDotSpacing,
        );
      case IndicatorType.expandingDots:
        return ExpandingDotsEffect(
          spacing: widget.indicatorDotSpacing,
          radius: widget.indicatorDotRadius,
          offset: widget.indicatorOffset,
          activeDotColor: widget.indicatorActiveDotColor,
          dotHeight: widget.indicatorDotHeight,
          dotWidth: widget.indicatorDotWidth,
          dotColor: widget.indicatorInActiveDotColor,
          strokeWidth: widget.indicatorStrokeWidth,
          paintStyle: widget.indicatorPaintStyle,
          expansionFactor: widget.indicatorExpansionFactor,
        );
      case IndicatorType.jumpingDots:
        return JumpingDotEffect(
          spacing: widget.indicatorDotSpacing,
          paintStyle: widget.indicatorPaintStyle,
          strokeWidth: widget.indicatorStrokeWidth,
          dotColor: widget.indicatorInActiveDotColor,
          dotWidth: widget.indicatorDotWidth,
          dotHeight: widget.indicatorDotHeight,
          activeDotColor: widget.indicatorActiveDotColor,
          offset: widget.indicatorOffset,
          radius: widget.indicatorDotRadius,
          jumpScale: widget.indicatorJumpScale,
          verticalOffset: widget.indicatorVerticalOffset,
        );
      case IndicatorType.scrollingDots:
        return ScrollingDotsEffect(
          activeStrokeWidth: widget.indicatorActiveStrokeWidth,
          spacing: widget.indicatorDotSpacing,
          paintStyle: widget.indicatorPaintStyle,
          strokeWidth: widget.indicatorStrokeWidth,
          dotColor: widget.indicatorActiveDotColor,
          dotWidth: widget.indicatorDotWidth,
          dotHeight: widget.indicatorDotHeight,
          activeDotColor: widget.indicatorActiveDotColor,
          offset: widget.indicatorOffset,
          radius: widget.indicatorDotRadius,
          activeDotScale: widget.indicatorActiveDotScale,
          fixedCenter: widget.indicatorFixedCenter,
          maxVisibleDots: widget.indicatorMaxVisibleDots,
        );
      case IndicatorType.slideDots:
        return SlideEffect(
          radius: widget.indicatorDotRadius,
          offset: widget.indicatorOffset,
          activeDotColor: widget.indicatorActiveDotColor,
          dotHeight: widget.indicatorDotHeight,
          dotWidth: widget.indicatorDotWidth,
          dotColor: widget.indicatorActiveDotColor,
          strokeWidth: widget.indicatorStrokeWidth,
          paintStyle: widget.indicatorPaintStyle,
          spacing: widget.indicatorDotSpacing,
        );
      case IndicatorType.scaleDots:
        return ScaleEffect(
          radius: widget.indicatorDotRadius,
          offset: widget.indicatorOffset,
          activeDotColor: widget.indicatorActiveDotColor,
          dotHeight: widget.indicatorDotHeight,
          dotWidth: widget.indicatorDotWidth,
          dotColor: widget.indicatorActiveDotColor,
          strokeWidth: widget.indicatorStrokeWidth,
          paintStyle: widget.indicatorPaintStyle,
          spacing: widget.indicatorDotSpacing,
          activePaintStyle: widget.indicatorActivePaintStyle,
          activeStrokeWidth: widget.indicatorActiveStrokeWidth,
          scale: widget.indicatorScale,
        );
      case IndicatorType.swapDots:
        return SwapEffect(
          spacing: widget.indicatorDotSpacing,
          radius: widget.indicatorDotRadius,
          offset: widget.indicatorOffset,
          activeDotColor: widget.indicatorActiveDotColor,
          dotHeight: widget.indicatorDotHeight,
          dotWidth: widget.indicatorDotWidth,
          type: widget.indicatorSwapType,
          dotColor: widget.indicatorInActiveDotColor,
          strokeWidth: widget.indicatorStrokeWidth,
          paintStyle: widget.indicatorPaintStyle,
        );
      case IndicatorType.colourTransitionDots:
        return ColorTransitionEffect(
          paintStyle: widget.indicatorPaintStyle,
          strokeWidth: widget.indicatorStrokeWidth,
          dotColor: widget.indicatorInActiveDotColor,
          dotWidth: widget.indicatorDotWidth,
          dotHeight: widget.indicatorDotHeight,
          activeDotColor: widget.indicatorActiveDotColor,
          offset: widget.indicatorOffset,
          radius: widget.indicatorDotRadius,
          spacing: widget.indicatorDotSpacing,
          activeStrokeWidth: widget.indicatorActiveStrokeWidth,
        );
      case IndicatorType.customizableIndicator:
        return CustomizableEffect(
          dotDecoration: widget.indicatorInActiveDotDecoration ??
              const DotDecoration(
                borderRadius: BorderRadius.all(Radius.circular(2.0)),
              ),
          activeDotDecoration: widget.indicatorActiveDotDecoration ??
              const DotDecoration(
                borderRadius: BorderRadius.all(Radius.circular(10.0)),
              ),
          spacing: widget.indicatorDotSpacing,
          inActiveColorOverride: widget.indicatorInActiveColorOverride,
          activeColorOverride: widget.indicatorActiveColorOverride,
        );
    }
  }
}

/// enum [IndicatorType] for selecting the type of page indicator.
enum IndicatorType {
  worm,
  expandingDots,
  jumpingDots,
  scrollingDots,
  slideDots,
  scaleDots,
  swapDots,
  colourTransitionDots,
  customizableIndicator,
}

/// enum [IndicatorPosition] for selecting the position of page indicator.
enum IndicatorPosition {
  center,
  topCenter,
  topRight,
  topLeft,
  bottomCenter,
  bottomRight,
  bottomLeft,
  centerRight,
  centerLeft,
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:google_fonts/google_fonts.dart';
import 'package:onboarding_animation/onboarding_animation.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'OnBoarding Animation',
      home: OnBoardingAnimationExample(),
    );
  }
}

class OnBoardingAnimationExample extends StatefulWidget {
  const OnBoardingAnimationExample({Key? key}) : super(key: key);

  @override
  State<OnBoardingAnimationExample> createState() =>
      _OnBoardingAnimationExampleState();
}

class _OnBoardingAnimationExampleState
    extends State<OnBoardingAnimationExample> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.white.withOpacity(.9),
      body: Padding(
        padding: const EdgeInsets.all(8.0),
        child: OnBoardingAnimation(
          controller: PageController(initialPage: 1),
          pages: const [
            _GetCardsContent(
              image: 'assets/images/image_1.jpg',
              cardContent:
                  'Lorem Ipsum is simply dummy text of the printing and typesetting industry.',
            ),
            _GetCardsContent(
              image: 'assets/images/image_2.jpg',
              cardContent:
                  'Lorem Ipsum is simply dummy text of the printing and typesetting industry.',
            ),
            _GetCardsContent(
              image: 'assets/images/image_3.jpg',
              cardContent:
                  'Lorem Ipsum is simply dummy text of the printing and typesetting industry.',
            ),
          ],
          indicatorDotHeight: 7.0,
          indicatorDotWidth: 7.0,
          indicatorType: IndicatorType.expandingDots,
          indicatorPosition: IndicatorPosition.bottomCenter,
          indicatorSwapType: SwapType.normal,
        ),
      ),
    );
  }
}

class _GetCardsContent extends StatelessWidget {
  final String image, cardContent;

  const _GetCardsContent({
    Key? key,
    this.image = '',
    this.cardContent = '',
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      decoration: const BoxDecoration(
        color: Colors.white,
        borderRadius: BorderRadius.all(
          Radius.circular(20.0),
        ),
      ),
      child: Padding(
        padding: const EdgeInsets.all(11.0),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          children: [
            ClipRRect(
              borderRadius: const BorderRadius.all(
                Radius.circular(20.0),
              ),
              child: Image.asset(image),
            ),
            Text(
              cardContent,
              style: GoogleFonts.pacifico(fontSize: 30.0),
            ),
          ],
        ),
      ),
    );
  }
}
